### PR TITLE
[dbnode] Ensure index data consistency 

### PR DESCRIPTION
**Pull request by notbdu**
**What this PR does / why we need it**:

Ensure index data consistency by persisting TSDB data after persisting index data during a cold flush. 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note
NONE
```

### CODE REVIEW

File: src/dbnode/persist/fs/merger.go [hunk 1]
```diff
@@ -178,7 +178,8 @@ func (m *merger) Merge(
 			nsCtx.Schema,
 			encoderPool)
 	)
-	defer func() {
+	// This gets called later when we finish persist work.
+	cleanupFn := func() {
```
Comment (arnikola) [inline src/dbnode/persist/fs/merger.go:182]: Shouldn't this still get called if there's an error encountered?

File: src/dbnode/persist/fs/merger_test.go [hunk 2]
```diff
@@ -463,8 +467,9 @@ func testMergeWith(
 		BlockStart: startTime,
 	}
 	mergeWith := mockMergeWithFromData(t, ctrl, diskData, mergeTargetData)
-	err := merger.Merge(fsID, mergeWith, 1, preparer, nsCtx, &persist.NoOpColdFlushNamespace{})
+	prepared, err := merger.Merge(fsID, mergeWith, 1, preparer, nsCtx, &persist.NoOpColdFlushNamespace{})
 	require.NoError(t, err)
+	require.NoError(t, prepared.Close())
```
Comment (arnikola) [inline src/dbnode/persist/fs/merger_test.go:472]: nit: worth adding a tracker to the test DeferClose  method to verify it's called correctly?

File: src/dbnode/persist/fs/write.go [hunk 3]
```diff
@@ -327,13 +329,41 @@ func (w *writer) Close() error {
...
+		w.digestFdWithDigestContents.Digest().Sum32(),
+		w.digestBuf,
+	); err != nil {
 		w.err = err
 		return err
 	}
 	return nil
 }
 
+func (w *writer) DeferClose(cleanup func()) (persist.DeferredCloser, error) {
```
Comment (robskillington) [inline src/dbnode/persist/fs/write.go:343]: We actually don't need to pass in a cleanup function here.

Writing the checkpoint file doesn't require keeping IDs or tags in memory, all the other files are written by the time we only have the checkpoint file to write left.

File: src/dbnode/persist/types.go [hunk 4]
```diff
@@ -40,10 +40,19 @@ type DataFn func(id ident.ID, tags ident.Tags, segment ts.Segment, checksum uint
 // blocks for a (shard, blockStart) combination.
 type DataCloser func() error
 
+// DeferCloser returns a DeferredCloser that persists the data checkpoint file when called.
+type DeferCloser func(cleanup func()) (DeferredCloser, error)
```
Comment (robskillington) [inline src/dbnode/persist/types.go:44]: Can we just make this:
```go
type DeferCloser func() (DataCloser, error)
```

So we can reuse the same type signature between the previous "Close" and the new "Close"?

File: src/dbnode/persist/fs/merger.go (hunk 1 above)
Comment (robskillington) [reply src/dbnode/persist/fs/merger.go:182]: If we change the deferred close to not take a cleanup fn (since writing checkpoint file doesn't depend on these resources) we can revert this just to a straight up "defer" again.

File: src/dbnode/storage/namespace.go [hunk 5]
```diff
@@ -1211,6 +1215,9 @@ func (n *dbNamespace) ColdFlush(flushPersist persist.FlushPreparer) error {
 	if onColdFlushDone != nil {
 		multiErr = multiErr.Add(onColdFlushDone())
 	}
+	for _, shardColdFlush := range shardColdFlushes {
+		multiErr = multiErr.Add(shardColdFlush.Done())
```
Comment (robskillington) [inline src/dbnode/storage/namespace.go:1219]: Need to make sure we only call this if no other errors yeah?

i.e.
```
for _, shardColdFlush := range shardColdFlushes {
  if multiErr.NumErrors() != 0 {
    continue
  }
  multiErr = multiErr.Add(shardColdFlush.Done())
}
```

Probably want to do the same with evicting/rotating in-memory index results:
```go
if multiErr.NumErrors() == 0 {
  if err := onColdFlushNs.Done(); err != nil {
    multiErr = multiErr.Add(err)
  }
}
if multiErr.NumErrors() == 0 && onColdFlushDone != nil  {
  multiErr = multiErr.Add(onColdFlushDone())
}
```

File: src/dbnode/persist/fs/merger_test.go [hunk 6]
```diff
@@ -450,7 +451,12 @@ func testMergeWith(
 				})
 				return nil
 			},
-			Close: func() error { return nil },
+			DeferClose: func() (persist.DataCloser, error) {
+				return func() error {
+					deferClosed = true
```
Comment (arnikola) [inline src/dbnode/persist/fs/merger_test.go:456]: nit: worth adding `require.False(t, deferClose)` to ensure we don't double close?

File: src/dbnode/persist/fs/merger_test.go [hunk 7]
```diff
@@ -463,8 +469,10 @@ func testMergeWith(
 		BlockStart: startTime,
 	}
 	mergeWith := mockMergeWithFromData(t, ctrl, diskData, mergeTargetData)
-	err := merger.Merge(fsID, mergeWith, 1, preparer, nsCtx, &persist.NoOpColdFlushNamespace{})
+	close, err := merger.Merge(fsID, mergeWith, 1, preparer, nsCtx, &persist.NoOpColdFlushNamespace{})
 	require.NoError(t, err)
+	require.NoError(t, close())
```
Comment (arnikola) [inline src/dbnode/persist/fs/merger_test.go:474]: nit: check `require.False(t, deferClosed)` before this to ensure we're not closing too early?

File: src/dbnode/persist/fs/write.go [hunk 8]
```diff
@@ -327,13 +329,38 @@ func (w *writer) Close() error {
...
+		w.digestBuf,
+	); err != nil {
 		w.err = err
 		return err
 	}
 	return nil
 }
 
+func (w *writer) DeferClose() (persist.DataCloser, error) {
+	err := w.close()
```
Comment (arnikola) [inline src/dbnode/persist/fs/write.go:344]: Worth erroring if this is called on a closed writer?

File: src/dbnode/persist/fs/write.go [hunk 9]
```diff
@@ -327,13 +329,38 @@ func (w *writer) Close() error {
...
+	}
+	if err != nil {
+		w.err = err
+		return nil, err
+	}
+	checkpointFilePath := w.checkpointFilePath
+	digestBuf := w.digestBuf
+	digestChecksum := w.digestFdWithDigestContents.Digest().Sum32()
+	return func() error {
+		return w.writeCheckpointFile(
```
Comment (arnikola) [inline src/dbnode/persist/fs/write.go:356]: Are there any synchronization concerns here? May be worth adding a comment to the interface about usage?

File: src/dbnode/storage/namespace.go [hunk 10]
```diff
@@ -1196,21 +1196,33 @@ func (n *dbNamespace) ColdFlush(flushPersist persist.FlushPreparer) error {
...
+		if err := onColdFlushNs.Done(); err != nil {
+			multiErr = multiErr.Add(err)
+		}
 	}
-	if onColdFlushDone != nil {
+	if multiErr.NumErrors() == 0 && onColdFlushDone != nil {
 		multiErr = multiErr.Add(onColdFlushDone())
 	}
+	for _, shardColdFlush := range shardColdFlushes {
+		if multiErr.NumErrors() != 0 {
```
Comment (arnikola) [inline src/dbnode/storage/namespace.go:1221]: nit; you can do this check outside of the loop

File: src/dbnode/storage/namespace.go [hunk 11]
```diff
@@ -1196,21 +1196,33 @@ func (n *dbNamespace) ColdFlush(flushPersist persist.FlushPreparer) error {
 		return err
 	}
 
+	// NB(bodu): Deferred shard cold flushes so that we can ensure that cold flush index data is
```
Comment (arnikola) [inline src/dbnode/storage/namespace.go:1199]: Not part of this PR, but noticed that `n.metrics.flushColdData.ReportSuccessOrError(res, n.nowFn().Sub(callStart))` won't get called if we hit one of the errors above; is this correct behaviour?

File: src/dbnode/persist/fs/merger.go [hunk 12]
```diff
@@ -89,7 +89,7 @@ func (m *merger) Merge(
 	flushPreparer persist.FlushPreparer,
 	nsCtx namespace.Context,
 	onFlush persist.OnFlushSeries,
-) (err error) {
+) (closer persist.DataCloser, err error) {
```
Comment (robskillington) [inline src/dbnode/persist/fs/merger.go:92]: nit: Can we avoid using named returns? Looks like they're not being used (no "return" statements) and it's too easy to over shadow accidently.

Maybe best at top of func to just declare `var closer persist.DataCloser`?

File: src/dbnode/persist/fs/merger_test.go (hunk 6 above)
Comment (robskillington) [reply src/dbnode/persist/fs/merger_test.go:456]: +1

File: src/dbnode/persist/fs/merger_test.go (hunk 7 above)
Comment (robskillington) [reply src/dbnode/persist/fs/merger_test.go:474]: +1

File: src/dbnode/persist/fs/write.go (hunk 9 above)
Comment (robskillington) [reply src/dbnode/persist/fs/write.go:356]: This whole structure is meant to be called in a single threaded way.

I do think though it is reused, so perhaps we shouldn't be using the the `w.digestBuf` (which seems to be reused?)

Might be best to even just to do entirely inline without shared resources and create a new stateless "func writeCheckpointFile(path string, checksum uint32) error" like so:

```go
 func writeCheckpointFile(filePath string, checksum uint32, newFileMode os.FileMode) error {
  fd, err := OpenWritable(filePath, newFileMode)
  // handle err ...
  err := digest.NewBuffer().WriteDigestToFile(fd, checksum)
  // handle err ...
  return fd.Close()
}
```

File: src/dbnode/storage/namespace.go (hunk 11 above)
Comment (robskillington) [reply src/dbnode/storage/namespace.go:1199]: Hm, it should since "res = multiError.FinalError()"?

File: src/dbnode/storage/namespace.go (hunk 10 above)
Comment (robskillington) [reply src/dbnode/storage/namespace.go:1221]: Yeah true, I guess we probably want to try to call "all dones" if we can.

File: src/dbnode/storage/namespace.go (hunk 11 above)
Comment (arnikola) [reply src/dbnode/storage/namespace.go:1199]: Yeah I mean in the error paths above this

File: src/dbnode/storage/namespace.go (hunk 10 above)
Comment (arnikola) [reply src/dbnode/storage/namespace.go:1221]: Ah good point; then I think there's a bug here, since it'll continue as soon as the first one fails

Comment (notbdu) [reply src/dbnode/storage/namespace.go:1221]: Then what should the expected behaviour be? We can't roll back a cold flush for already flushed shards and the logic previously was basically the same. Attempt to flush each shard and log errors for failed shards.

File: src/dbnode/persist/fs/write.go (hunk 8 above)
Comment (notbdu) [reply src/dbnode/persist/fs/write.go:344]: There's currently no mechanism tracking open/closed state and the upstream caller calls open/close in a single thread. I think it's fine to leave the expectations as is.

File: src/dbnode/storage/namespace.go (hunk 10 above)
Comment (arnikola) [reply src/dbnode/storage/namespace.go:1224]: I think expected behaviour is to call done on each flush, and just gather all the errors?
```golang
for _, shardColdFlush := range shardColdFlushes {
 multiErr = multiErr.Add(shardColdFlush.Done())
}
```

Comment (notbdu) [reply src/dbnode/storage/namespace.go:1224]: Hmm, I'll move the error check around. I think we need to check the error still before attempting to complete shard cold flushes. Since if the index cold flusher fails, we don't want to complete the shard cold flush since those series are not persisted in the index yet.

File: src/dbnode/persist/fs/write.go [hunk 13]
```diff
@@ -327,13 +329,39 @@ func (w *writer) Close() error {
...
+		return nil, err
+	}
+	checkpointFilePath := w.checkpointFilePath
+	digestChecksum := w.digestFdWithDigestContents.Digest().Sum32()
+	return func() error {
+		return writeCheckpointFile(
+			checkpointFilePath,
+			digestChecksum,
+			digest.NewBuffer(),
+			w.newFileMode,
```
Comment (robskillington) [inline src/dbnode/persist/fs/write.go:360]: Maybe also take ref to `w.newFileMode` above declaring the lambda? If for whatever reason that changed between reuse?

File: src/dbnode/storage/namespace.go [hunk 14]
```diff
@@ -1173,6 +1173,7 @@ func (n *dbNamespace) ColdFlush(flushPersist persist.FlushPreparer) error {
 
 	resources, err := newColdFlushReuseableResources(n.opts)
 	if err != nil {
+		n.metrics.flushColdData.ReportError(n.nowFn().Sub(callStart))
```
Comment (arnikola) [inline src/dbnode/storage/namespace.go:1176]: 👍

File: src/dbnode/storage/namespace.go [hunk 15]
```diff
@@ -1187,37 +1188,48 @@ func (n *dbNamespace) ColdFlush(flushPersist persist.FlushPreparer) error {
...
-		multiErr = multiErr.Add(err)
+	if multiErr.NumErrors() == 0 {
+		if err := onColdFlushNs.Done(); err != nil {
+			multiErr = multiErr.Add(err)
+		}
 	}
-	if onColdFlushDone != nil {
-		multiErr = multiErr.Add(onColdFlushDone())
+	if multiErr.NumErrors() == 0 && onColdFlushDone != nil {
+		if err := onColdFlushDone(); err == nil {
```
Comment (arnikola) [inline src/dbnode/storage/namespace.go:1221]: nit: flip to reduce nesting?
